py-polars/src/dataframe/import.rs

```diff
@@ -0,0 +1,93 @@
+use polars::export::arrow::ffi::{ArrowArrayStream, ArrowArrayStreamReader};
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::types::{PyCapsule, PyType};
+
+use super::*;
+
+/// Validate PyCapsule has provided name
+fn validate_pycapsule_name(capsule: &Bound<PyCapsule>, expected_name: &str) -> PyResult<()> {
+    let capsule_name = capsule.name()?;
+    if let Some(capsule_name) = capsule_name {
+        let capsule_name = capsule_name.to_str()?;
+        if capsule_name != expected_name {
+            return Err(PyValueError::new_err(format!(
+                "Expected name '{}' in PyCapsule, instead got '{}'",
+                expected_name, capsule_name
+            )));
+        }
+    } else {
+        return Err(PyValueError::new_err(
+            "Expected schema PyCapsule to have name set.",
+        ));
+    }
+
+    Ok(())
+}
+
+/// Import `__arrow_c_stream__` across Python boundary.
+fn call_arrow_c_stream<'py>(ob: &'py Bound<PyAny>) -> PyResult<Bound<'py, PyCapsule>> {
+    if !ob.hasattr("__arrow_c_stream__")? {
+        return Err(PyValueError::new_err(
+            "Expected an object with dunder __arrow_c_stream__",
+        ));
+    }
+
+    let capsule = ob.getattr("__arrow_c_stream__")?.call0()?.downcast_into()?;
+    Ok(capsule)
+}
+
+pub(crate) fn import_stream_pycapsule(capsule: &Bound<PyCapsule>) -> PyResult<PyDataFrame> {
+    validate_pycapsule_name(capsule, "arrow_array_stream")?;
+
+    // Takes ownership of the pointed to ArrowArrayStream
+    // This acts to move the data out of the capsule pointer, setting the release callback to NULL
+    let stream_ptr =
+        Box::new(unsafe { std::ptr::replace(capsule.pointer() as _, ArrowArrayStream::empty()) });
```

Comment:
This is the same thing that arrow-rs does [here](https://github.com/apache/arrow-rs/blob/9be0eb530de85a68358dc31c146f71ccb97739ca/arrow-array/src/ffi_stream.rs#L177-L191)